tests/chainer_tests/dataset_tests/test_convert.py

```diff
@@ -0,0 +1,197 @@
+
+    def check_concat_arrays(self, arrays, device=None):
+        array = dataset.concat_examples(arrays, device)
+        self.assertEqual(array.shape, (len(arrays),) + arrays[0].shape)
+
+        for x, y in zip(array, arrays):
+            numpy.testing.assert_array_equal(
+                cuda.to_cpu(x), cuda.to_cpu(y))
+
+    def test_concat_arrays_cpu(self):
+        arrays = self.get_arrays_to_concat(numpy)
+        self.check_concat_arrays(arrays)
+
+    @attr.gpu
+    def test_concat_arrays_gpu(self):
+        arrays = self.get_arrays_to_concat(cuda.cupy)
+        self.check_concat_arrays(arrays)
+
+    @attr.gpu
+    def test_concat_arrays_to_gpu(self):
+        arrays = self.get_arrays_to_concat(numpy)
+        self.check_concat_arrays(arrays, cuda.Device().id)
+
+    def get_tuple_arrays_to_concat(self, xp):
+        return [(xp.random.rand(2, 3), xp.random.rand(3, 4))
+                for _ in range(5)]
+
+    def check_concat_tuples(self, tuples, device=None):
+        arrays = dataset.concat_examples(tuples, device)
+        self.assertEqual(len(arrays), len(tuples[0]))
+        for i in range(len(arrays)):
+            shape = (len(tuples),) + tuples[0][i].shape
+            self.assertEqual(arrays[i].shape, shape)
+
+            for x, y in zip(arrays[i], tuples):
+                numpy.testing.assert_array_equal(
+                    cuda.to_cpu(x), cuda.to_cpu(y[i]))
+
+    def test_concat_tuples_cpu(self):
+        tuples = self.get_tuple_arrays_to_concat(numpy)
+        self.check_concat_tuples(tuples)
+
+    @attr.gpu
+    def test_concat_tuples_gpu(self):
+        tuples = self.get_tuple_arrays_to_concat(cuda.cupy)
+        self.check_concat_tuples(tuples)
+
+    @attr.gpu
+    def test_concat_tuples_to_gpu(self):
+        tuples = self.get_tuple_arrays_to_concat(numpy)
+        self.check_concat_tuples(tuples, cuda.Device().id)
+
+    def get_dict_arrays_to_concat(self, xp):
+        return [{'x': xp.random.rand(2, 3), 'y': xp.random.rand(3, 4)}
+                for _ in range(5)]
+
+    def check_concat_dicts(self, dicts, device=None):
+        arrays = dataset.concat_examples(dicts, device)
+        self.assertEqual(frozenset(arrays.keys()), frozenset(dicts[0].keys()))
+        for key in arrays:
+            shape = (len(dicts),) + dicts[0][key].shape
+            self.assertEqual(arrays[key].shape, shape)
+
+            for x, y in zip(arrays[key], dicts):
+                numpy.testing.assert_array_equal(
+                    cuda.to_cpu(x), cuda.to_cpu(y[key]))
+
+    def test_concat_dicts_cpu(self):
+        dicts = self.get_dict_arrays_to_concat(numpy)
+        self.check_concat_dicts(dicts)
+
+    @attr.gpu
+    def test_concat_dicts_gpu(self):
+        dicts = self.get_dict_arrays_to_concat(cuda.cupy)
+        self.check_concat_dicts(dicts)
+
+    @attr.gpu
+    def test_concat_dicts_to_gpu(self):
+        dicts = self.get_dict_arrays_to_concat(numpy)
+        self.check_concat_dicts(dicts, cuda.Device().id)
+
```

Comment:
I usually make one test fixture for one feature of the library, and separate the test cases for different usages, thus I made one test class in this case. Maybe it is a sign of bad design if you feel the test fixture should be divided into several ones for only one functionality (I mean, it would be better to separate the `concat_examples` into multiple different functions, though I think the current design is much more Pythonic)...